Add nondeterministic tags in tags.yaml and add the nondeterministic_seeded tag to all functions in native_functions.yaml defined as nondeterministic by alias_analysis.cpp

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81440

- This PR adds the nondeterministic tag to tags.yaml to specify functions that may not necessarily return the same outputs when ran with identical inputs. 
- The tag is added to the functions in native_functions.yaml that are specified as nondeterministic by aliasdb in https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.cpp#L1146
- **Thus there may be ops that are nondeterministic that currently do not have the nondeterministic tag but should. The plan is to create a test bench to determine which ops in native_functions.yaml are nondeterministic and add the tag to qualifying functions in a later pr.**

Commit:
[JIT] Add nondeterministic tag in tags.yaml and add the tag to all functions in native_functions.yaml defined as nondeterministic by alias_analysis.cpp

[ghstack-poisoned]

anjali411 reviewed (commented):
This PR looks good overall and it'd be cool for users to be able to query this information in Python. Although, I am curious to hear about your motivating use case to add this tag?

We have a testing setup to test tags for all our OpInfos, see here: 
https://github.com/pytorch/pytorch/blob/8175a8bcb535d094077393f8185ee725fdf0ff61/test/test_ops.py#L1467-L1495 It should be easy to add a test for this tag!

goldenxuett commented:
> This PR looks good overall and it'd be cool for users to be able to query this information in Python. Although, I am curious to hear about your motivating use case to add this tag?
> 
> We have a testing setup to test tags for all our OpInfos, see here:
> 
> https://github.com/pytorch/pytorch/blob/8175a8bcb535d094077393f8185ee725fdf0ff61/test/test_ops.py#L1467-L1495
> 
> It should be easy to add a test for this tag!

The motivating use case is that I'm creating a class that consolidates this information (amongst various other information) from sources that were originally baked into JIT. I was recommended to use tags rather than copying the hard coded list from AliasDB. This will also let me remove the List from AliasDB as the class I'm creating will be used by AliasDB.

Some other questions:
1. Since there are failing checks on this test, is there anything I need to override to make them pass? Or can I merge with failing changes.
It seems like it is necessary to fail the tests in order to add the change due to the nature of the changes.

2. Speaking of merging, can I merge this PR the same way as I have my other PRs (ie: telling pytorchbot to merge) or is there some other commands I should run.

anjali411 commented:
> The motivating use case is that I'm creating a class that consolidates this information (amongst various other information) from sources that were originally baked into JIT. I was recommended to use tags rather than copying the hard coded list from AliasDB. This will also let me remove the List from AliasDB as the class I'm creating will be used by AliasDB.

yup that sounds great!

> Some other questions:
> 
> 1. Since there are failing checks on this test, is there anything I need to override to make them pass? Or can I merge with failing changes.

you should add the tag name to this list https://github.com/pytorch/pytorch/blob/cce2f0d0e48a0fff85cf9d445b14b2ab9f050cc7/test/test_public_bindings.py#L33-L249 and that should fix the test failures.

> 2. Speaking of merging, can I merge this PR the same way as I have my other PRs (ie: telling pytorchbot to merge) or is there some other commands I should run.

Yeah you can merge the PR in that way once it is approved.

As per my previous comment, it'd be nice to add a test for this new tag. It'd also be a quick way to ensure that all the relevant ops are labeled with this tag. Let me know if you have any questions!

goldenxuett commented:
> As per my previous comment, it'd be nice to add a test for this new tag. It'd also be a quick way to ensure that all the relevant ops are labeled with this tag. Let me know if you have any questions!

My plan was to add the test in a later pr after I figure out the best way to test for nondeterminism. 
I would want to ship this current pr with just the definitions as defined by AliasDB in order to finish up integrating my intern project with AliasDB, so that I can switch my attention fully back to this and other stretch goals. 

Is that fine?

anjali411 commented:
> > As per my previous comment, it'd be nice to add a test for this new tag. It'd also be a quick way to ensure that all the relevant ops are labeled with this tag. Let me know if you have any questions!
> 
> My plan was to add the test in a later pr after I figure out the best way to test for nondeterminism. I would want to ship this current pr with just the definitions as defined by AliasDB in order to finish up integrating my intern project with AliasDB, so that I can switch my attention fully back to this and other stretch goals.
> 
> Is that fine?

Sure

davidberard98 commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -16,3 +16,8 @@
   desc: |
           This tag indicates that the operator doesn't have an explicit entry in
           native_functions.yaml, and instead was generated automatically by the codegen.
+- tag: nondeterministic
+  desc: |
+          This tag indicates that the operator is nondeterministic meaning that if the
+          operator is run twice with the same arguments, the outputs will not necessarily be
+          identical.
```
we were discussing that we want to differentiate between:
a) numerical nondeterminism, for ops with implementations that don't guarantee bitwise equivalence across runs 
b) "actual" nondetermninism, for ops that intentionally produce different results, e.g. random ops.

and the intent of this tag was to only cover (b). Can you clarify this in the description?

cc @bdhirsh

bdhirsh replied:
Naming bikeshed for the two categories above

I'm not sure what the best name for those two categories should be, but some ideas to bikeshed on:

(a) bitwise_nondeterministic
(b) nondeterministic

(a) nondeterministic
(b) from_distribution

(a) nondeterministic
(b) random

@albanD @anjali411 any thoughts / ideas / suggestions on names? Since once we pick one, we're stuck with it :)

anjali411 replied:
I think `seeded_nondeterministic` would be clearer (since the ops in question are not truly nondeterministic -- they generate the same output for a given seed). cc. @pbelevich who previously worked on RNG 

As for `bitwise_nondeterminism`, it sometimes only applies for an op on certain backends I think (cc. @ngimel). So, idk if we want to add a tag for that or not.

bdhirsh replied:
I tentatively vote to swap the wording to `nondeterministic_seeded` and `nondeterminstic_bitwise`, so users are forced to see both and decide which one they want.

Commit:
Update on "[JIT] Add nondeterministic tag in tags.yaml and add the tag to all functions in native_functions.yaml defined as nondeterministic by AliasDB"


- This PR adds the nondeterministic tag to tags.yaml to specify functions that may not necessarily return the same outputs when ran with identical inputs. 
- The tag is added to the functions in native_functions.yaml that are specified as nondeterministic by aliasdb in https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.cpp#L1146
- **Thus there may be ops that are nondeterministic that currently do not have the nondeterministic tag but should. The plan is to create a test bench to determine which ops in native_functions.yaml are nondeterministic and add the tag to qualifying functions in a later pr.**

[ghstack-poisoned]

Commit:
Update on "[JIT] Add nondeterministic tag in tags.yaml and add the tag to all functions in native_functions.yaml defined as nondeterministic by AliasDB"


- This PR adds the nondeterministic tag to tags.yaml to specify functions that may not necessarily return the same outputs when ran with identical inputs. 
- The tag is added to the functions in native_functions.yaml that are specified as nondeterministic by aliasdb in https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.cpp#L1146
- **Thus there may be ops that are nondeterministic that currently do not have the nondeterministic tag but should. The plan is to create a test bench to determine which ops in native_functions.yaml are nondeterministic and add the tag to qualifying functions in a later pr.**

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -16,3 +16,11 @@
+  desc: |
+          This tag indicates if an operator is nondeterminstically seeded (ie is random)
+          such that the operator intentially produces different results when run twice on the same inputs.
+- tag: nondeterministic_bitwise
```
Hmm we shouldn't add this tag unless there's a plan to add this tag to the relevant ops

ngimel replied:
We have `set_nondeterministic_mode` that warns or errors out on bitwise-nondeterministic ops, and it's getting a lot of use, so, if going forward tagging is a common mechanism to classify ops, having `nondeterministic_bitwise` will be useful.

goldenxuett commented:
@pytorchbot merge
